pop_tools/xgcm_util.py

```diff
@@ -0,0 +1,169 @@
+    return ds_new
+
+
+def relabel_pop_dims(ds):
+    """Return a new xarray dataset with distinct dimensions for variables at different
+    grid points.
+    """
+    ds_new = _label_coord_grid_locs(ds)
+    ds_new = _add_pop_dims_to_dataset(ds_new)
+    for vname in ds_new.variables:
+        if 'grid_loc' in ds_new[vname].attrs:
+            da = ds_new[vname]
+            dims_orig = da.dims
+            new_spatial_dims = _dims_from_grid_loc(da.attrs['grid_loc'])
+            if dims_orig[0] == 'time':
+                dims = ('time',) + new_spatial_dims
+            else:
+                dims = new_spatial_dims
+            ds_new[vname] = xr.Variable(dims, da.data, da.attrs, da.encoding, fastpath=True)
+    return ds_new
+
+
+def get_xgcm_grid(ds, **kwargs):
+    """Return an xgcm Grid object
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+        An xarray Dataset
+    kwargs:
+       Additional keyword arguments are passed through to `xgcm.Grid` class.
+
+    Returns
+    -------
+    grid : xgcm.Grid
+        An xgcm Grid object
+
+    Examples
+    --------
+    >>> from pop_tools import get_xgcm_grid, DATASETS
+    >>> import xarray as xr
+    >>> fname = DATASETS.fetch("iron_tracer.nc")
+    >>> ds = xr.open_dataset(fname)
+    >>> ds
+    <xarray.Dataset>
+    Dimensions:    (nlat: 384, nlon: 320, time: 24, z_t: 60, z_w_bot: 60, z_w_top: 60)
+    Coordinates:
+    * time       (time) object 0249-02-01 00:00:00 ... 0251-01-01 00:00:00
+    * z_t        (z_t) float32 500.0 1500.0 2500.0 ... 487508.34 512502.8 537500.0
+        TLAT       (nlat, nlon) float64 ...
+        ULONG      (nlat, nlon) float64 ...
+        TLONG      (nlat, nlon) float64 ...
+        ULAT       (nlat, nlon) float64 ...
+    * z_w_top    (z_w_top) float32 0.0 1000.0 2000.0 ... 500004.7 525000.94
+    * z_w_bot    (z_w_bot) float32 1000.0 2000.0 3000.0 ... 525000.94 549999.06
+    Dimensions without coordinates: nlat, nlon
+    Data variables:
+        UE         (time, z_t, nlat, nlon) float32 ...
+        VN         (time, z_t, nlat, nlon) float32 ...
+        WT         (time, z_w_top, nlat, nlon) float32 ...
+        HDIFE      (time, z_t, nlat, nlon) float32 ...
+        HDIFN      (time, z_t, nlat, nlon) float32 ...
+        HDIFB      (time, z_w_bot, nlat, nlon) float32 ...
+        DIA_IMPVF  (time, z_w_bot, nlat, nlon) float32 ...
+        KPP_SRC    (time, z_t, nlat, nlon) float32 ...
+        STF        (time, nlat, nlon) float32 ...
+        SMS        (time, nlat, nlon) float32 ...
+    >>> grid = get_xgcm_grid(ds)
+    >>> grid
+    <xgcm.Grid>
+    Z Axis (periodic):
+        * center   z_t --> left
+        * right    z_w_bot --> center
+        * left     z_w_top --> center
+    Y Axis (periodic):
+        * center   nlat_t --> right
+        * right    nlat_u --> center
+    X Axis (periodic):
+        * center   nlon_t --> right
+        * right    nlon_u --> center
+
+    """
+    import xgcm
+
+    ds = relabel_pop_dims(ds)
```

Comment:
Fairly certain this doesn't work. `ds` isn't returned but a copy is made deep down in the call stack and all the xgcm stuff nrequires the renamed dimensions.

How about

`grid, new_ds = pop_tools.to_xgcm_grid_dataset(ds)`

Reply:
> How about

> `grid, new_ds = pop_tools.to_xgcm_grid_dataset(ds)`

👍 Good catch. I've addressed this and renamed the utility function